KIPs/kip-81.md

```diff
@@ -0,0 +1,783 @@
+        bool executed;          // true if successfully execute()d
+    }
+    mapping(uint256 => Proposal) proposals;
+
+    function state(uint256 proposalId) external view returns (ProposalState) {
+        Proposal storage p = proposals[proposalId];
+
+        if (p.executed) {
+            return ProposalState.Executed;
+        } else if (p.canceled) {
+            return ProposalState.Canceled;
+        } else if (block.number < p.voteStart) {
+            return ProposalState.Pending;
+        } else if (block.number <= p.voteEnd) {
+            return ProposalState.Active;
+        } else if (!checkQuorum(proposalId)) {
+            return ProposalState.Failed;
+        }
+
+        if (!p.queued) {
+            if (block.number <= p.queueDeadline) {
+                return ProposalState.Passed;
+            } else {
+                return ProposalState.Expired;
+            }
+        } else {
+            if (block.number <= p.execDeadline) {
+                return ProposalState.Queued;
+            } else {
+                return ProposalState.Expired;
+            }
+        }
+    }
+}
+```
+
+#### Executing transactions
+
+The transactions in a passed proposal are executed in a order they appear in the `propose()` function arguments.  Each transaction is sent to `targets[i]`, carrying `values[i]` of KLAYs, with `calldatas[i]` as input data. Below is an example of `execute()` function.
+
+```solidity
+abstract contract Voting {
+    struct Proposal {
+        address[] targets;
+        uint256[] values;
+        bytes[] calldatas;
+
+    }
+    mapping(uint256 => Proposal) proposals;
+
+    function execute(uint256 proposalId) external {
+        Proposal storage p = proposals[proposalId];
+
+        for (uint256 i = 0; i < p.targets.length; i++) {
+            (bool success, bytes memory result) =
+                    p.targets[i].call{value: p.values[i]}(p.calldatas[i]);
+            require(success, "Transaction failed");
+        }
+    }
+}
+```
+
+## Rationale
+
+#### CnStakingV2 and StakingTracker instead of standard ERC20Votes
+
+It is common for DAOs to manage their governance tokens in a [ERC20Votes](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20Votes) contract. However new contracts CnStakingV2 and StakingTracker were used to manage staked KLAYs. Using the new contracts provides better compatibility with ecosystem Dapps like public staking services. It also does not break the existing block validator selection algorithm that depends on AddressBook.
+
+#### StakingTracker refreshStake and refreshVoting has no access control
+
+Another desion options was to restrict those funtions to only CnStakingV2 contracts. But those functions are publicly open to allow GCs to gradually migrate from CnStakingContract to CnStakingV2 contracts. Old CnStakingContract does not notify StakingTracker about its balance change, in which case manually calling `refreshStake()` can notify the change.
```

Comment:
@blukat29 
Can I understand this like:
Even if a GC member wants to use CnStakingContract, then they can use it with manual call of refreshStake() of StakingTracker. Is this right? no other side effects expected?

Reply:
Yes, you are right. I designed it so that staking contracts can be optionally & gradually upgraded.

If a GC member uses old CnStakingContract,
- Do not call `refreshStake()` at all -> The initial balance at `createTracker()`, or `propose()` is used.
- Do call `refreshStake()` -> The balance change during the preparation period is recognized.

Calling `refreshStake()` only affects the preparation period. I guess most GC members don't frequently change their stake amounts, thus there is little need to call it. However, they can choose to utilize the preparation period.